Use queue to manage controllers and add timing settings per controller

TD-er opened the pull request:
In short, I did a lot of refactoring to remove a lot of code duplication in the controllers.

All MQTT controllers now use a delay queue and the other controllers use their own queue, since the data per controller request may differ per controller.
This queue does allow some pacing of the data to the receiving end and some buffering to make sure sending is non blocking.
It may also help to send the controller pushes generated when the wifi was not yet available.  (important for deep sleep use cases)

Things that can be set per controller:
![image](https://user-images.githubusercontent.com/3751318/44346593-a11ecc80-a496-11e8-8fa4-693e07924d1f.png)

**Max Queue Depth** is the number of items allowed in the queue.
Since this may take some memory, try to keep this low. Maybe the current limit of 10 is a bit too low?

**Minimum Send Interval** is the minimal time between two controller send actions.
This will mainly be a limiting factor when sending in a burst, since the last send time of that specific controller is taken into account.

**Max Retries** is the maximum number of retries to send a queued value.
For some controllers, which send per plugin value, this value must be at least higher than the number of items per plugin which has to be sent.

**Full Queue Action** is about the used strategy when sending the queue is full.
For example for Thingspeak the send interval is high, so you may want to set the queue length to 1 and this strategy value to "Delete Oldest" so you will always get the last value.

Also a lot of code duplication in the HTTP based controllers is moved to some special functions.
This saves a lot of code and makes it easier to fix issues or add features of which all plugins can benefit.


N.B.
At this moment only controller C001 ... C008 have this queue implemented.
The rest may experience some issues since the MessageDelay variable is now not being used anymore for sending tasks.

TD-er committed:
[controller delay] Let controllers handle their own delay and buffer

TD-er committed:
[MQTT] Add controller delay queue for MQTT

MQTT messages are queued and processed at set pace to not overload the broker.

N.B. other controllers still need a similar queue to process. Current interval for those is now probably too low.

Also initial queue parameters must still be corrected when set to 0.

TD-er committed:
[Controller] Make ControllerDelayHandler templated and comment

Added recognisable comments throughout the `ESPEasy-Globals.h` file and made ControllerDelayHandler templated to be used for other controllers to use.

TD-er committed:
[C001] Let C001 controller use queue

TD-er committed:
[C003] Let C003 controller use queue

TD-er committed:
[C004] Let C004 controller use queue and simplify using macros

Added some simplifications to the code by creating one big macro in `_CPlugin_Helper.h` that does all the work and defines.
This will reduce the chance of typo errors a lot.

TD-er committed:
Added `USES_Cxxx` checks

functions may not be mentioned in the scheduler when the controller is not included in the build.

TD-er committed:
[C007] Let C007 controller use queue

TD-er committed:
[MQTT] Use set controller settings on MQTT controllers

TD-er committed:
[Webserver] Add fields for new Controller settings

Make sure all new settings are configurable for all controllers.

TD-er committed:
[C008] Let C008 use queue and some bugfixes

Also added statistics
And some optimizations

TD-er committed:
[#1666] Moved code duplication for HTTP requests to special functions

The controller source files are now much cleaner and all use the same code.
So it is easier to fix issues and add features and checks.

TD-er commented:
Would be nice if people could already test this.
I will make a test build for it as soon as Travis is happy :)

TD-er commented:
My laptop had some (heat) issues building it, but I managed to build a number of versions.
[ESPeasy_PullReq_1669_controller-1-8.rar](https://www.dropbox.com/s/lw643fbvt37ub3v/ESPeasy_PullReq_1669_controller-1-8.rar?dl=0)

clumsy-stefan commented:
@TD-er: I'm happy to test this (in my own build), however, I only use controller 009 (FHEM), so I'm not sure if this helps in any way?

on the other hand it would be great if you could update C009 also with the new queues?!

clumsy-stefan commented:
does not work for me, authentication fails for C009 (fhem) controller after this update, even after updating the credentionals and setup of the controller!

TD-er commented:
OK, tomorrow evening I can have a look at C009. It is one I didn't have a loot at yet (and may have touched a little to make it get rid of the code duplication)
At least good to know it is now broken :)

micropet commented:
@ TD-er

I just flashed it. What should I pay attention to?
I use openHAB MQTT

Grovkillen commented:
@micropet if it still works then it's great :) Other than that you may see in the MQTT log (listen to that node's publishing topics) and look for any abnormal things. I have tested the openHAB for some hours and that one worked good for me. I don't use an actual controller on the receiving end though, only my nodes are using MQTT as communication amongst themselves.

TD-er commented:
One of the things you could look for is if it remains responsive and stable when sending a lot of messages.
Also since it is now using a queue to keep the messages, you may also see samples being sent when the wifi is lost or recorded samples from before the wifi was connected. (at boot for example)
You can also play with the settings at controller level to see if these all work like expected.
For example setting the delay between messages higher or lower.

It should no longer be blocking and work better. At least that was my intention ;)

micropet commented:
Yes OK. Runs now 9 hours without problems with the following settings:

Minimum Send Interval: 100 [ms]
Max Queue Depth: 10  
Max Retries: |10
Full Queue Action: Ignore new

Annotation:

The versions of the last months do not run a full day. Every few hours comes a watchdog reboot.

But one thing has gotten better: After the reboot, the devices are always back on the net and reachable.

![device_page_1](https://user-images.githubusercontent.com/2838584/44385801-d1ab4880-a520-11e8-889f-c8c6c5179a3d.png)

TD-er commented:
Yep, it still puzzles me what is causing these watchdog reboots.
I have experienced it only once on one of my nodes (at least that I am aware of) and that node is already running for over 2.5 days now since the last time it got a new version.
N.B. it isn't running this PR yet.

micropet commented:
After flashing 9 hours without reboot.
Then reboot every 2-3 hours with Hardware Watchdog.

MQTT is OK.

TD-er commented:
What may have changed?
More traffic on the network?
Or maybe less so the access point is in sleep mode?

micropet commented:
I think it's a coincidence.
Currently there are 51 devices in the WLAN. The WLAN runs over two access points from ubiquitte and is extremely stable.

The last few months happened every few hours. a watchdog reboot
This happens even if no plugins are loaded.

But as I said: The Nodes then go back to the WLAN and continue to work.

TD-er committed:
[C009] Removed port number from HTTP host header

As suggested here: https://github.com/letscontrolit/ESPEasy/pull/1669#issuecomment-414434578
This plugin is not using the Queue yet, but only had some duplicate code moved to a helper class.
I now removed the port number from the HTTP header, since I am not really sure if the standard demands it or only for the connect part.
Also changed the HTTP command to use the code that has the port in the header, just to check if that will break now compared to older builds.

TD-er commented:
@clumsy-stefan I just pushed a small fix for C009.
I looked into what I changed in that plugin and the only thing I could find was I now add the port to the HTTP header when it is a non-standard port.
See my description here: https://github.com/letscontrolit/ESPEasy/pull/1669/commits/82134961729117bf220a7839f24f1561fa23b840

Could you please test if C009 now works again?
And maybe someone could test the HTTP command (`sendtohttp` via rules) between this last patch and some older build to a non-standard port (other than 80) to see if it now works different from before.
The HTTP command now uses the code that inserts a port number in the "host" part of the HTTP header.

TD-er committed:
[C009] Controller C009 now also uses Delay Queue

Also fixed an issue where sending some value to a controller would fail when the buffer is full and the settings set to "Delete Oldest".

clumsy-stefan commented:
@TD-er: tried it, still does not work... on the receiving end it tells me "Basic authentication active but no credentials received" so I assume something in the HTTP (header) that provides the HTTP authentication is not working anymore...reverting back to the mega commit (without this PR) works again!

TD-er commented:
That is helpful information.
I will look into it.
Thanks for testing

TD-er committed:
[Client Connect] Check for success on connecting with client.

In case of domain name resolution error result can be negative.
See [Dns.cpp](https://github.com/esp8266/Arduino/blob/18f643c7e2d6a0da9d26ff2b14c94e6536ab78c1/libraries/Ethernet/src/Dns.cpp#L44)

Thus must match the result with 1.

TD-er committed:
[C009] Remove 'no port' option in HTTP packet

As tested by @clumsy-stefan, it didn't make a difference.

TD-er committed:
[HTTP Auth] Now using proper index for retrieving credentials

I was using the controller number and not the controller index to get the user credentials.
Should now work for FHEM.

TD-er commented:
@clumsy-stefan It should now work with the credentials in FHEM.
But this update has the status "it compiles".
Took me a while to find what I was doing wrong, so I had changed and reverted a lot of code and this is the moment to go to bed and let others decide whether it does indeed work (controllers 1 ... 9)
Tomorrow I will do some  testing.

clumsy-stefan commented:
@TD-er: great work!!! thank you very much, and sorry for keeping you up late!!!

did a build this morning including all commits from this PR and it seems to run fine!! I'll keep it running now and report on any issues!

clumsy-stefan commented:
my units seem to run stable until now (at least as stable as before)!! so good work!!

micropet commented:
My units are the same as always.
I do not notice a difference.

clumsy-stefan commented:
just realised another thing: if you have lots of tasks sending messsages (eg. 12 tasks every 60 secs.) it seems that the max queue depth of 10 can get full too fast and some of the values are never beeing transferred (my guess)... . I'll try to increase queuedepth definition and see if it resolves the issue!

TD-er commented:
It is limited in the webserver side.
But I guess you're right. Maybe we should set the limit to something like 25, or maybe some memory limit on the buffer?
Those elements for the controller can get quite large for some controllers.

clumsy-stefan commented:
changed it in the webserver to 20 (max) and set it in the interface to 20 then. this at least solves my issue.

so I suggest to either raise the limit, make it configurable but at least check for enough free mem left..

TD-er committed:
[HTTP] Add user agent in request header

Should make this optional later

TD-er committed:
[C010] Controller C010 now also uses Delay Queue

TD-er committed:
[Cleanup] Merge code duplication between C008 and C010

Both use the same Queue Element class.

TD-er committed:
[C011] Controller C011 now also uses Delay Queue

And some code cleanup of shared code in C001 and C003.
Also fixed an issue in C011, where the result was not returned when sending a value.

TD-er commented:
Added controller C010 and C011.
Would be nice if someone can test these.

Also if someone could try and build these using the Arduino IDE.
I use some macros in here, of which I hope the Arduino IDE does also compile it well.

TD-er committed:
[Controller] Moved the 'magic' default and max values to defines

The Max and Default values for the Controller Delay Queue are now set in the `ESPEasy-Globals.h` file.

micropet commented:
Arduino does not work:

WARNUNG: Kategorie 'LED' in der Bibliothek CHT16K33 LED Matrix Library ist ungültig und wird auf 'Uncategorized' festgelegt
WARNUNG: Kategorie 'Input' in der Bibliothek MechInputs ist ungültig und wird auf 'Uncategorized' festgelegt
WARNUNG: Kategorie 'Input' in der Bibliothek SerialSensors ist ungültig und wird auf 'Uncategorized' festgelegt
WARNUNG: Bibliothek LiquidCrystal_I2C behauptet auf (avr) Architektur(en) ausgeführt werden zu können und ist möglicherweise inkompatibel mit Ihrem derzeitigen Board, welches auf (esp8266) Architektur(en) ausgeführt wird.
_P016_IR:11:1: error: 'IRrecv' does not name a type

 IRrecv *irReceiver;
 ^
_P016_IR:12:1: error: 'decode_results' does not name a type
 decode_results results;
 ^
D:\arduino-1.8.3\meine\ESPEasy-mega-08-28\ESPEasy\_P016_IR.ino: In function 'boolean Plugin_016(byte, EventStruct*, String&)':

_P016_IR:56:13: error: 'irReceiver' was not declared in this scope
         if (irReceiver == 0 && irPin != -1)
             ^
_P016_IR:59:27: error: expected type-specifier before 'IRrecv'
           irReceiver= new IRrecv(irPin);
                           ^
_P016_IR:59:27: error: expected ';' before 'IRrecv'

_P016_IR:62:13: error: 'irReceiver' was not declared in this scope

         if (irReceiver != 0 && irPin == -1)
             ^
_P016_IR:66:18: error: type '<type error>' argument given to 'delete', expected pointer

........

TD-er commented:
If the only errors are in P016, then I guess it is just an issue with switching branches and the library location for the Arduino IDE.
Then I'm not worried.
I will rebase this branch on top of Mega, to make switching easier.

Edit: @9:26 Done rebase

micropet commented:
OK.

TD-er committed:
[C012] Controller C012 now also uses Delay Queue

TD-er committed:
[C013] Remove some obsolete function

`C013_Send` did actually only call another function and some other bogus object loading without using the data.

TD-er committed:
[Nodes] Make Nodes list a map with dynamic size

Nodes list is now a std::map, which may contain up-to 254 nodes (1...254), since 255 is reserved for broadcast and 0 for "all known".

TD-er commented:
And again some new commits added.
C012 is now also using delay queue.
As preparation for C013, I changed the Nodes array with fixed length to a std::map
Nodes can have a unit number ranging from 1...254.
Actually having so many nodes will impose some issues on memory use, but that's for later.

clumsy-stefan commented:
merged the latest commits and rebuilt/reflashed my units... enabled udp discovery (inter ESP networking) on some nodes to try.

how often should this list get updated respectively how often do the nodes send theys presence out? after about 15min. there is still only the local node in the nodelist.

can't test the other controllers though, I only use C009..

TD-er commented:
After enabling the udp port you still have to reboot the first time.
All nodes should share the same UDP port.

clumsy-stefan commented:
yep, that was it, I was not aware that after enabling the port I need to reboot.... now all configured units are showing up... I'll test them and will feedback if I encounter any issues... so far everything looks good though..

TD-er commented:
The best reports are of-course "I see no difference" and an occasional "seems to be working more smooth" :)

clumsy-stefan commented:
Sure, just didn't want to spam you/here with "everything looks good" too much ;)

but anyways, currently the updated nodes run stable for about 3h... not a lot but a good start ;)

TD-er committed:
[Nodes] Highlight current node in the list of nodes

TD-er committed:
[Cleanup] Memory optimisation in available controllers and plugins

These structs to keep track of available plugins and controllers were fixed in size and also the member variables were not ordered in the most optimum way.
Also the name of the nodes always used 26 Bytes and would impose a memory leak when deleted.

clumsy-stefan commented:
All nodes ran fine until this morning (~20h)... 

The only thing I see is that the CPU load isquite high (at least after startup) however the webinterface does respond swiftly.... so probably a calculation issue with the new queueing mechanism?

I included the latest commits now and reflashed the units as well as enabled interESP udp discovery again for all nodes (~40)... let's see what happens ;)

micropet commented:
Hi clumsy-stefan.

Can you provide the source?
I can not get it all together.

clumsy-stefan commented:
what source?
(I pulled it from GIT and compiled myself)

micropet commented:
The last Source with all changes from TD-er.

clumsy-stefan commented:
I have it from here? you can just do a git pull and merge this PR in... You would need all my local configs to compile my version, which does anyways only include a subset of plugins (the ones I need)... so I'm sorry, but I think this won't work...

micropet commented:
Then I'm too stupid to find it.

Good, then I'm still waiting.

clumsy-stefan commented:
sorry, I think I misunderstood... the changes from @TD-er are all within this pull-request #1669. I only used these changes, no other ones.
if you update your local repository and merge this pull request onto your local repository ou should have all the latest changes?!
Are you using gitkraken or a similar gui-tool?

micropet commented:
I do not use a git tool. I download the zip file with the source and compile it. :)
I do not know how to merge something.
But no matter. I'll wait until TD-he integrates it.

clumsy-stefan commented:
ahh, sorry ;) you're right, that's a bit more difficult then to use the latest test-code...
as I use special defines (in custom.h) that correspond to the Arduino IDE I'm using I think just "sending" the source probably won't work... sorry for this...

micropet commented:
Is no problem.

I have holidays today and wanted to spend a new version on my units.
The versions of the last few weeks run unfortunately only a few hours, then comes the watchdog reboot.
Because you wrote something of 20 hours, I wanted to give it a try.

clumsy-stefan commented:
@TD-er: I had to disable the UDP discovery again. it somehow seems like the nodes get too busy listening and working on all the UDP packets if you have lots of nodes... they get unresponsive quite quickly (sorry, I was wrong with my first statement, that they stay responsive)...
after disabling also CPU usage returned to a "normal" percentage again..

TD-er commented:
@micropet Is this the one you can use to test?
[PR1669_test_ESP8266_4096_VCC.rar](https://www.dropbox.com/s/a7ryva6pb8zax9s/PR1669_test_ESP8266_4096_VCC.rar?dl=0)

micropet commented:
Fine. I just flashed it.
I do not need VCC - but does not seem to bother.

TD-er commented:
@micropet I will try to setup a build-script for me to automatically generate test builds given an url.
This makes it easier to build pre-release test versions.

micropet commented:
Yes, that's a very good idea.

TD-er committed:
[Nodes] Fix String copy to properly read node names

TD-er committed:
[MQTT] Give MQTT state in log when connect state changes.

TD-er committed:
[SPIFFS] Add SPIFFS size + usage to sysinfo page

micropet commented:
The version does not run for me 20 hours.
Every 3 to 5 hours the comes the watchdog reset.

Testet on one Unit without Sensors and two Units with Sensors.

TD-er commented:
OK, thanks for the update.
I really need to look into what's causing these watchdog resets, since I get more reports (even without these changes) and I have really not a clue where to look for it.

micropet commented:
I also think that it is difficult.
The problem has been around for a few months.

It is not critical because the units re-enter the network after the reboot.

There were versions last year where the units no longer log into the network. That was extremely bad.

TD-er commented:
Apart from the reboots, does this (massive) code change work like expected?
And what controllers did you use?
Have you also changed some of the new settings in the controller?

micropet commented:
This version is otherwise "normal". :)
That means she works as always.

I only use the MQTT controller with
Minimum Send Interval: 100
Max Cue Depth: 10
Max Retries: 10
Full Queue Action: Ignore New

I do not know what to change, all measurements are transmitted as desired via MQTT.

TD-er commented:
The controllers now use a queue, with these settings you gave.
This means the recorded values will no longer get lost when wifi is not (yet) connected.

Also sending the values is less blocking, so it should react more 'snappy'.

clumsy-stefan commented:
not as a qualified report, but I have the feeling when my server is loaded and probably doesn't answer in due time it seems that the units now get first more load (CPU >90%) and then reboots at some point in time... does not seem a memory leak though as I record memory and CPU usage and except high CPU load I can't see memory decreasing...

one thin I can imagine is that if the server is slow (eg. accepts the connection but can't receive data) the number of open connections on the node get too high... because this is what happens sometimes on my server..

before the controller used the queue it just rebooted at some point without this visible increase in CPU load...

I use sonoff basic, pow and 4ch's as well as wemos mini and pro's.... self compiled version including this PR and FHEM as server and controller.

TD-er commented:
I was already looking into the way the watchdog timers are being fed.
And I will also look into async like communications to add some timeout feature.

TD-er committed:
[Watchdog] Add watchdog feed to backgroundtasks() function

A lot of reports recently about resets by the Hardware Watchdog.
So maybe this timer is not reset as often as before.

TD-er commented:
Could you please test [this build](https://www.dropbox.com/s/ak20lfmv06natjc/ESPEasy_pr1669-2-g14816bd.zip?dl=0) ?
It does feed the software watchdog timer and I hope that also calls the hardware watchdog timer.

clumsy-stefan commented:
I also just built a new version with 14816bd included and currently flashing my nodes... However I had uptimes between 200 and 1'500 minutes before flashing, so it's really different from unit to unit... even exactly equal configured unit reboot completely independent from each other, that's why I think it's related to the network/server comms...

micropet commented:
@TD-er flasht your build. Will see...

giig1967g commented:
I am experiencing the hardware watchdog issue too.
I have two different locations where I use the ESP8266.
In one location I have an ASUS router and things seems to go smooth.
In another location I have a Linksys router (previously Mikrotik) and I am having the HW WD.

Will try your latest build and report.

clumsy-stefan commented:
I still do have some random HW-Watchdog reboots... uptime between 200 and 500min. currently

micropet commented:
2 Units without reboot since 7h
1 unit Manual reboot (2) Reset Reason | Hardware Watchdog

TD-er committed:
[MQTT] Force MQTT client reconnect when wifi reconnects

It looks like the check for `client->connected()` may return false positives.
This means the MQTT client may assume the connection is still active, while it isn't.

TD-er commented:
[Build with last commit](https://www.dropbox.com/s/3pmi25omsybvo99/ESPEasy_pr1669-3-g7f7837f.zip?dl=0)

micropet commented:
After 3 h the first Unit enters Hardware Watchdog.
2 Units running since 4h

TD-er commented:
OK, so it isn't a proper fix. At best it will make it less reproducible. :(

micropet commented:
Jo. :)
But... two are running.

giig1967g commented:
My two units have been going for 23h without rebooting.
Will keep you posted.

TD-er commented:
Also when they do reboot due to hardware watchdog, please also give an indication on reboot frequency before this patch was applied.

micropet commented:
So I can not specify a reboot frequency.
At the moment I have only 15 units in operation.
I can only say that they boot several times a day. (2 - 5 times)
Running > 1 day is rare.

TD-er commented:
Do you have the wifi bandwidth set to 40 MHz? or 20 MHz?

micropet commented:
On 2.4 GHz, i have 20 MHz.

giig1967g commented:
results so far:
1 unit is running for 2 days and 15 hour without issues.

Another unit crashed after 2 days and 10 hours with following error:
```
Boot | Manual reboot (6)
Reset Reason | Exception

```

Differences between the two:
first unit has one PCF8574 for output and internal GPIO for input.
second unit has two PCF8574, one for input and one for output.

TD-er commented:
@giig1967g Do you also have a log/chart of the memory usage on the nodes?
The exception can also occur if a memory allocation fails.

giig1967g commented:
@TD-er : I don't have it.
How can I do it?

micropet commented:
ESP-201:
Uptime | 0 days 2 hours 15 minutes
Load | 34.00% (LC=15397)
Free Mem | 18768 (15016 - ruleMatch2)
Boot | Manual reboot (7)
Reset Reason | Hardware Watchdog

ESP-206:
Uptime | 0 days 1 hours 23 minutes
Load | 28.30% (LC=18106)
Free Mem | 18800 (16536 - ruleMatch2)
Boot | Manual reboot (7)
Reset Reason | Hardware Watchdog

ESP-218:
Uptime | 2 days 1 hours 58 minutes
Load | 10.40% (LC=20968)
Free Mem | 20208 (17912 - parseTemplate3)
Boot | Cold boot (0)
Reset Reason | Software/System restart

![image](https://user-images.githubusercontent.com/2838584/45027468-c1778b00-b041-11e8-9d40-1cba324ab572.png)

![image](https://user-images.githubusercontent.com/2838584/45027513-e4a23a80-b041-11e8-8117-cef4b0d31528.png)

![image](https://user-images.githubusercontent.com/2838584/45027631-3ba80f80-b042-11e8-97fc-38f1c57a2ee3.png)

giig1967g commented:
Another difference between the two units is that the one that crashed with exception has 4 pages of rules, while the one that is still alive has only 2.
I don't know if this could make the difference

TD-er commented:
@giig1967g You could check a few times on that node to see what is the lowest recorded free mem and what function was being called when that lower limit was reached.
The functions that parse rules may take some memory. I see those functions also on my nodes every now and then.

giig1967g commented:
Unit that crashed:  Free Mem | 18608 (14440 - ruleMatch2)
Unit that didn't crash (yet - so far 3d 4h) : Free Mem | 18312 (14552 - sendContentBlocking)

giig1967g commented:
One thing that just happened in one node (the one that didn't crash yet) is:
it just disconnected and immediately reconnected from MQTT while I was browsing in the user interface looking for the system info. Wifi remain connected.

TD-er committed:
[WDT] Add yield() and runbackgroundtasks() when parsing rules

Rules parsing may take some time, so perform some yields to prevent watchdog triggers.

TD-er commented:
I also made a [test build for the latest commit](https://www.dropbox.com/s/j9j5sc2xjw5ba0l/ESPEasy_pr1669-4-ge6105bf.zip?dl=0)
Filenames in the zip have again a rather dated looking filename, but that's something in my build script and `git describe`.

giig1967g commented:
Update on the unit that crashed.
Uploaded new test build. 
Free RAM = 19552 (11168 - sendContentBlocking)

TD-er committed:
[#1707] Getting logs from the /log even when password is not entered

See #1707

TD-er committed:
[#1714] Parse notification sender to allow some sender formatting

`bla@bla.com` will be set as `nodename <bla@bla.com>` in the mail header as from part.
`John Doe <bla@bla.com>` will now be used like that in the mail header.  

N.B. `<` is actually the separator, so as long as that character is being used to separate the name and mail address, it will work. (see #1714 )

TD-er commented:
I know C013 is not included yet, but do the other ones already implemented feel OK enough to merge tomorrow?

giig1967g commented:
@TD-er: I have found a solution for #1709.
Can you verify and merge it together with this PR?

TD-er commented:
I commented on your suggestion and whatever comes out of that, will be merged with this PR.

clumsy-stefan commented:
@TD-er: from my point of view this PR could be merged. especially e6105bf seems to improve stability again! (have it running for 24h hours without significant issues, even though some units rebooted)

something I still expierience is when the controller can connect to the server but not get rid of it's data (eg. tcp connection connected, but the receiver end does not accept data) the unit's go to >90% CPU and I guess at some point then the watchdog kicks in... need to verify though...

EDIT: PS: probably yeld() and runbackgroundtasks() or simply update the watchdog should be somewhere in the data-sending part of the controller too?

giig1967g commented:
For me the stability is ok.
1 unit has been running for 4 days and 16 hours.
My 2nd unit has been running without rebooting since the very latest commit update (1 day).

clumsy-stefan commented:
PS: I think e6105bf and following commits have not been included in this PR ? I fetched them from your repository...

TD-er commented:
@clumsy-stefan They appear to be included in the list of commits at the top of this PR.

clumsy-stefan commented:
strange, GITKraken doesn't show it as a commit in this branch/PR.... but if it's included that's fine with me ;)

TD-er commented:
Maybe you should do a fetch?
I have done a few times a rebase to the mega branch to make sure this isn't getting too much out of sync.

clumsy-stefan commented:
no, but also  in this pull request on gitweb it does not show as a commit... see first screenshot, it does not show as a commit, as in the second screenshot all the older commits...

<img width="882" alt="image" src="https://user-images.githubusercontent.com/32835100/45176959-e02c7c00-b211-11e8-9908-c4c9890896c7.png">

<img width="889" alt="image" src="https://user-images.githubusercontent.com/32835100/45177001-fafef080-b211-11e8-8af2-fd7736c7ec19.png">

TD-er commented:
That list is from the moment I made a new rebase.

clumsy-stefan commented:
that's just what I see in here... but again, if it's included once you include this PR it into mega, I guess all's fine then... especially because this one commit made it a lot more stable for me ;)

TD-er committed:
[Scheduler] Do not schedule task device timer for tasks without interval

See #1709

TD-er commented:
For those interested, I made a test build including the last commit added.
[test build](https://www.dropbox.com/s/wg3zyb6iih0zhbk/ESPEasy_pr1669-5-g4e95bbe.zip?dl=0)

For info on what had changed, see: https://github.com/letscontrolit/ESPEasy/issues/1709#issuecomment-419235717

giig1967g commented:
tested latest build.
It fixes one of the issues in #1709.

clumsy-stefan commented:
did also a fresh build including the last 3 commits and flashed this morning.. testing now... 

still trying to find out, why I don't see this commits as part of this PR...

micropet commented:
Flasht it before 4h20mins

Uptime | 0 days 2 hours 12 minutes
Load | 33.50% (LC=11290)
Free Mem | 15256 (12720 - parseTemplate3)
Boot | Manual reboot (3)

clumsy-stefan commented:
had 2 out of 25 untis rebooting until now (~3hours since flashing them):

Local Time | 2018-09-07 10:07:47
Uptime | 0 days 0 hours 5 minutes
Load | 28.50% (LC=8009)
Free Mem | 16296 (920 - LoadControllerSettings)
Boot | Manual reboot (1)
Reset Reason | Hardware Watchdog


Local Time | 2018-09-07 10:10:47
Uptime | 0 days 0 hours 59 minutes
Load | 35.80% (LC=7053)
Free Mem | 15152 (1304 - LoadControllerSettings)
Boot | Manual reboot (1)
Reset Reason | Hardware Watchdog

TD-er commented:
Hmm that memory usage while in `LoadControllerSettings` does worry me.

TD-er commented:
What controllers do you use and have you increased the queue length?
Maybe I should add some protection in there to virtually limit the queue when memory is getting full.

The process delay queue function does call LoadControllerSettings , so it may be at the lowest point of free memory.

TD-er committed:
[Queue] Add memory check to determine full queue

Depending on the type of controller, the queued messages can be quite big.
This means the queue limit set can cause the memory to fill up.
This commit adds another check. If the free memory is less than 5000 bytes, it will consider the queue being full.
It will then also log the actual memory usage of the queue.

TD-er commented:
I will merge this one and implementation of C013 and other changes will be done in new issues/commits.

clumsy-stefan commented:
I'm unsing controller 009 (FHEM). sorry have troubles with my server currently, can't really test... but I think that's exaclty one of the issues, as then the queues get full...

clumsy-stefan commented:
just saw a bugfix in WifiCLient.ino example in the esp8266 core library which deals with slow servers:
` while (client.available()) {`
becomes
` while (client.available() || client.connected()) {`

probably that's the issue I'm facing? Could/should that be included in ESPEasy too?

TD-er commented:
Hmm, that's an OR.
I will have a look at that. In which file is this change applied?.

Edit:
Never mind, WiFiClient.ino example you said :)
https://github.com/esp8266/Arduino/commit/4fdff072e836a03df5ccd3ddf98ff8ba39b29db8

TD-er commented:
@clumsy-stefan I am making a fix for that, but I was wondering.
Do you also get this in the logs? 

> "Timeout while reading input data!"

If so, then we should also add a parameter in the settings to tweak the timeout.

v-a-d-e-r commented:
I'm testing the new controller queue version. One small thing could be fixed: When I update to the new version and a supported controller was already configured, the new fields show all '0'. There should be something like: if field = 0 (or -1) then field = default value (100, 10 ,10) for the first time. Otherwise ESPeasy is running mad, because in all 3 fields '0' is not allowed.. ;-)

micropet commented:
The current build does not run for me over one hour without reboot.

firmware
Build 20102 - Mega
Libraries ESP82xx Core 00000000, NONOS SDK 2.2.1 (cfd48f3), LWIP: 2.0.3
GIT version mega-20180910
Plugins 72 [Normal] [Testing]
Build Md5 182f11b4191da95d31374caefffca38

TD-er commented:
@micropet What controllers and plugins are active?
How much memory free?
What is the reboot reason?
And what version of the core libraries? 000000 isn't the 1.8.0 version.

TD-er commented:
@v-a-d-e-r When 0, I will add a patch to correct them to the defaults, which they already use to operate.

But now I come to think of it, I assume those new values are 0, when not set. I will also have a look if they are properly bound checked.

v-a-d-e-r commented:
Only to be sure: This happens only when you flash the new firmware over an old firmware without the queue and the controller is already set. When you setup a new controller, then the values are 100 ,10,10.

BTW: My 16 Nodes (all are different) with the new controller queue are running without an issue since over 24h now!

TD-er commented:
Ah, that's what I need, some good news :)
There is a versioning system built in the settings, maybe I can increase that one to force the new values to some proper defaults.

v-a-d-e-r commented:
In the 'web interface controller page' must somewhere checked if the current config file has already the new struct with the 3 new values. If not --> use default values

Yeah, good news are ever welcome. ;-) I hope that this controller queue stuff becomes not a twin of the 'event based WiFi' problems.... *grin*

sassod commented:
With today's release the blynk plugin stalls all the processes with "
Timeout while reading input data!"

If the blynk plugin is disabled all works ok, I did an upgrade to todays
release all the queue values were at 0 corrected them but still no success.
I use my own blynk server and it has been running stable for a couple of
months now. If someone wants to test it the ip is in the pictures and the
subscribe new user is opened with 100k energy.... 😁

![screenshot_20180910-220144](https://user-images.githubusercontent.com/3404543/45322649-299a0580-b549-11e8-979a-1ec6b79289f3.jpg)
![screenshot_20180910-215847](https://user-images.githubusercontent.com/3404543/45322650-2a329c00-b549-11e8-80b6-1ccdae1c69c7.jpg)



On Mon, Sep 10, 2018, 20:26 v-a-d-e-r <notifications@github.com> wrote:

> In the 'web interface controller page' must somewhere checked if the
> current config file has already the new struct with the 3 new values. If
> not --> use default values
>
> Yeah, good news are ever welcome. ;-) I hope that this controller queue
> stuff becomes not a twin of the 'event based WiFi' problems.... *grin*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/letscontrolit/ESPEasy/pull/1669#issuecomment-420012895>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADPy_29RsXDnQqv9CPVDFc5rcwWG8SnMks5uZq7GgaJpZM4WEJLo>
> .
>

micropet commented:
Until the version ESP_Easy_mega-20180909_test_ESP8266_4096.bin the units ran a few hours to reboot.

Version ESP_Easy_mega-20180910_test_ESP8266_4096.bin comes to no hour.

The units are mixed. All run with the Wemos D1.
Two have no hardware connected, the rest mostly: BME280 BH1750 MH-Z19 TVoc CSS811 SDS021 Neopixel

A few post higher I have three graphs showing the relationship between memory and boot.

The most Units have this configuration:

![device_page_1](https://user-images.githubusercontent.com/2838584/45339396-a8fff700-b592-11e8-8d7f-01444c20b80b.png)

TD-er commented:
What controller are you using?

micropet commented:
MQTT (openHAB)

giig1967g commented:
Testing latest version 20180910:
Unit has been up for 16 hours; now all of a sudden I am getting MQTT disconnections (4 times in few minutes). Wifi didn't disconnect. Unit did not reboot.

Memory usage is worring me:
Free Mem | 18920 (1104 - ruleMatch2)

Controller is OpenHab.

TD-er commented:
I get the feeling a lot of the issues we see are due to memory issues.
Either low on memory, or memory fragmentation.
Both may lead to allocation errors and those may lead to undefined behavior.
Also the MQTT disconnect issues are something others have reported over the last few months and not only on ESPeasy.
I need to do some proper checking if MQTT is still connected. Maybe there is an option to ask the broker if it thinks the client is still connected?

giig1967g commented:
In my case I see the folllowing errors in mosquitto logs on the broker:
```
1536652191: Client ESPT2_2 has exceeded timeout, disconnecting.
1536652191: Socket error on client ESPT2_2, disconnecting.
1536652192: New connection from 192.168.88.202 on port 1883.
1536652192: New client connected from 192.168.88.202 as ESPT2_2 (c1, k10, u'openhabian').
1536652218: Socket error on client ESPT2_2, disconnecting.
1536652219: New connection from 192.168.88.202 on port 1883.
1536652219: New client connected from 192.168.88.202 as ESPT2_2 (c1, k10, u'openhabian').
1536652234: Client ESPT2_2 has exceeded timeout, disconnecting.
1536652234: Socket error on client ESPT2_2, disconnecting.
1536652237: New connection from 192.168.88.202 on port 1883.
1536652237: New client connected from 192.168.88.202 as ESPT2_2 (c1, k10, u'openhabian').
1536652270: Socket error on client ESPT2_2, disconnecting.
1536652271: New connection from 192.168.88.202 on port 1883.
1536652271: New client connected from 192.168.88.202 as ESPT2_2 (c1, k10, u'openhabian').
1536652286: Client ESPT2_2 has exceeded timeout, disconnecting.
1536652286: Socket error on client ESPT2_2, disconnecting.
1536652287: New connection from 192.168.88.202 on port 1883.
1536652287: New client connected from 192.168.88.202 as ESPT2_2 (c1, k10, u'openhabian').

```

TD-er commented:
Numbers in seconds (unix time stamp) I assume?
And is it MQTT import, or a regular controller? (which one?)

giig1967g commented:
Yes, EPOCH timestamp.
Mine is regular MQTT controller connected to OpenHAB.
I am not using MQTT import at the moment.

uzi18 commented:
it looks like we need to use wireshark/tcpdump to find what is going on ...
